### PR TITLE
Optimize joins with similar subqueries

### PR DESCRIPTION
**Pull request by zhenxiao**
Follow
https://github.com/facebook/presto/issues/1732

A first step, optimize queries of the following shape:

SELECT * 
FROM (
   SELECT k, agg1, agg2
   FROM t
   GROUP BY k
) a
JOIN (
   SELECT k, agg3, agg4
   FROM t
   GROUP BY k
) b
ON (a.k = b.k)

as

SELECT k, agg1, agg2, agg3, agg4
FROM T
GROUP BY k

### CODE REVIEW

**Comment by zhenxiao** [comment]
@martint An updated version, build summaries for child, check optimize criteria, and finally build new plan using summaries
In the summaries, I kept some projection assignments, and when building the new plan, reuse them and merge them together to get the new plan
Your comments and suggestions are appreciated

**Comment by zhenxiao** [comment]
@martint Found UnaliasSymbolReferences will get rid of the intentionally put aliases, which is to preserve AggregationNode outputs for the upper level OutputNode
Do you think there is better way to not running UnaliasSymbolReferences after SimilarJoinSubqueryOptimizer? Or I just disable it if featuresConfig.isOptimizeSimilarJoins() is set

**Comment by zhenxiao** [comment]
@martint I explicitly disable UnaliasSymbolReferences after SimilarJoinSubqueryOptimizer in this version. Any comments are appreciated.

**Comment by martint** [comment]
You shouldn't have to disable that optimization. What I meant with "you need to preserve the ouput of the join" is that if you have a plan that looks like:

```
output -> join -> agg -> ...
              |-> agg -> ...
```

The rewrite needs to transform the join into something else. In order to make things simple, the new subplan needs to produce the same outputs as the join (so that the optimization rule doesn't need to worry about anything that happens above it).

It's fine for optimizations that run after to alter that shape. Assuming they are not buggy, they shouldn't change it in ways that make it incorrect.

**Comment by zhenxiao** [comment]
@martint thank you. get it updated. Your comments are appreciated

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 1]
```diff
@@ -97,4 +98,16 @@ public FeaturesConfig setOptimizeSingleDistinct(boolean optimizeSingleDistinct)
         this.optimizeSingleDistinct = optimizeSingleDistinct;
         return this;
     }
+
+    public boolean isOptimizeSimilarJoins()
+    {
+        return optimizeSimilarJoins;
+    }
+
+    @Config("optimizer.optimize-similar-joins")
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: A more accurate name would be "optimizer.optimize-joined-aggregations"

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 2]
```diff
@@ -1108,6 +1108,13 @@ public void testSimpleJoin()
     }
 
     @Test
+    public void testSimilarJoin()
+            throws Exception
+    {
+        assertQuery("SELECT * FROM (SELECT orderstatus, min(totalprice) FROM ORDERS GROUP BY orderstatus) a JOIN (SELECT orderstatus, max(custkey * 5) - min(orderkey * 10) FROM ORDERS GROUP BY orderstatus) b ON (a.orderstatus = b.orderstatus)");
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Don't capitalize `ORDERS`

**Comment by zhenxiao** [comment]
thanks @martint 
get comments addressed

**Comment by zhenxiao** [comment]
@martint kindly ping

**Comment by yuananf** [comment]
Hi, I'm trying to write an optimizer to fix #3364 , can you offer any advice? Thank you.

**Comment by zhenxiao** [comment]
Hi @yuananf I think Presto needs some table stats to be able to optimize join orders. Otherwise have to do hints(Hive has some hints about MapJoin, etc). Let's ask @martint for more advice.

**Comment by yuananf** [comment]
My case seems like part of CBO, and CBO needs a lot of work.
Presto needs to add CBO like this:
http://zh.hortonworks.com/blog/hive-0-14-cost-based-optimizer-cbo-technical-overview/

Anyway, thank you!

**Comment by zhenxiao** [comment]
@martint I rebased the code at:
https://github.com/facebook/presto/pull/3607
Your comments and suggestions are appreciated

**Pull request closed without merging**